transmission.go

```diff
@@ -145,68 +148,51 @@ func respondErr(e *encoder, handle uint64, err error) {
 	rep.encode(e)
 }
 
-// ctxRW wraps a net.Conn to provide cancellation. It does so by setting a
-// low-ish timeout on each read/write call. If the call times out, it checks if
-// its context is done and if not, retries it.
+// ctxRW wraps a net.Conn to respect context cancellation. It does so by
+// starting a goroutine that sets the connection's read/write deadline in the
+// past whenever the context is cancelled.
 type ctxRW struct {
-	ctx   context.Context
-	c     net.Conn
-	hasDL bool
-	dl    time.Time
+	ctx    context.Context
+	cancel context.CancelCauseFunc
+	c      net.Conn
 }
 
 // wrapConn wraps a connection in a ctxRW.
-func wrapConn(ctx context.Context, c net.Conn) io.ReadWriter {
-	dl, ok := ctx.Deadline()
-	return &ctxRW{ctx, c, ok, dl}
+func wrapConn(ctx context.Context, c net.Conn) io.ReadWriteCloser {
+	// Note: cancel is called by Close().
+	ctx, cancel := context.WithCancelCause(ctx)
+	go func() {
+		<-ctx.Done()
+		c.SetDeadline(time.Now())
```

Comment:
Hm there is still a very subtle race I'm a bit dissatisfied with. If code does

```go
func F(ctx context.Context, c net.Conn) {
    wc := wrapConn(ctx, c)
    wc.Close()
    c.Close()
}
```

The call to `c.Close()` is concurrent to the call to `c.SetDeadline`. *In practice* this shouldn't really cause any issues with "normal" `net.Conn` implementations, but it would feel cleaner to me, if `wc.Close()` would only return after this goroutine finishes.

A reasonably simple way to do that would be to add a `done <-chan struct{}` field to `ctxRW`, `defer close(done)` at the top of this function-literal and a `<-rw.done` before the return from `(*ctxRW).Close`.

Does that make sense?

Reply:
Ah yes, good catch. Done.